mission/mission_handler_landing.hxx

```diff
@@ -0,0 +1,193 @@
+{
+    bool success = true;
+
+    print_util_dbg_print("Automatic-landing: descent_to_small_altitude\r\n");
+    auto_landing_behavior_ = DESCENT_TO_SMALL_ALTITUDE;
+    state_.mav_mode_custom &= static_cast<Mav_mode::custom_mode_t>(0xFFFFFFE0);
+    state_.mav_mode_custom |= Mav_mode::CUST_DESCENT_TO_SMALL_ALTITUDE;
+    is_landed_ = false;
+
+    waypoint_ = wpt;
+
+    return success;
+}
+
+template <class T1, class T2>
+Mission_handler::update_status_t Mission_handler_landing<T1, T2>::update()
+{
+    /*****************************
+    Handle internal landing states 
+    *****************************/
+    // Determine if we should switch between the landing states
+    bool next_state = false;
+    if (auto_landing_behavior_ == DESCENT_TO_GND)
+    {
+        alt_lpf_ = LPF_gain_ * (alt_lpf_) + (1.0f - LPF_gain_) * ins_.position_lf()[Z];
+        if ((ins_.position_lf()[Z] > -0.1f) && (maths_f_abs(ins_.position_lf()[Z] - alt_lpf_) <= 0.2f))
+        {
+            next_state = true;
+        }
+    }
+    else if (auto_landing_behavior_ == DESCENT_TO_SMALL_ALTITUDE)
+    {
+        float z_lim = desc_to_ground_altitude_- desc_to_ground_range_;
+        if ((ins_.position_lf()[Z]) > z_lim)                                         // Drone is below desc2gnd waypoint (for cases when we start landing below it)
+        {
+            next_state = true;
+        }
+    }
+    
+    /*************
+    Set controller 
+    *************/
+    bool ret = false;
+    switch (auto_landing_behavior_)
+    {
+        case DESCENT_TO_SMALL_ALTITUDE:
+            ret = set_desc_to_small_alt_control_command();
+            break;
+
+        case DESCENT_TO_GND:
+            ret = set_desc_to_ground_control_command();
+            break;
+    }
+
+    // If we are switching between states, ... then switch
+    if (next_state)
+    {
+        switch (auto_landing_behavior_)
+        {
+            case DESCENT_TO_SMALL_ALTITUDE:
+                print_util_dbg_print("Automatic-landing: descent_to_GND\r\n");
+                auto_landing_behavior_ = DESCENT_TO_GND;
+                state_.mav_mode_custom &= static_cast<Mav_mode::custom_mode_t>(0xFFFFFFE0);
+                state_.mav_mode_custom |= Mav_mode::CUST_DESCENT_TO_GND;
+                alt_lpf_ = ins_.position_lf()[Z];
+                break;
+
+            case DESCENT_TO_GND:
+                print_util_dbg_print("Auto-landing: disarming motors \r\n");
+                auto_landing_behavior_ = DESCENT_TO_SMALL_ALTITUDE;
+                //Do not reset custom flag here, to be able to check after landing
+                // in case something went wrong. Is reset while arming
+                state_.set_armed(false);
+                state_.mav_state_ = MAV_STATE_STANDBY;
+                is_landed_ = true;
+                break;
+        }
+    }
+    
+
+    /********************
+    Determine status code 
+    ********************/
+    if (waypoint_.autocontinue() == 1 && is_landed_)
```

Comment:
When you land through the landing handler, the drone will disarm automatically, which means that  the next iteration, the drone will switch to standby.

Autocontinue is used to determine if the drone should start to handle the next waypoint. Because when we land through the button, I assumed that we want the mission to be over so I made the button not be autocontinue. That means that the handler will not change to the interrupted waypoint in the list and will just switch to standby later. The handlers don't have to finish, they can be interrupted.

If autocontinue is set to on, then the drone will switch to standby in two methods based on how the landing waypoint was created. If the landing was a part of the mission, it will just switch to standby. If the mission was interrupted through a button or something, it will setup the interrupted waypoint again and then overwrite it with standby next iteration (this due to the drone resuming the mission after an inserted waypoint has been completed).

I left the autocontinue in there as I could see the mission plan requiring more than one landing and takeoff (e.g. delivery of goods). In that case, you would have the drone land with autocontinue on into some new mission item such as release goods and then takeoff or something. Of course, the mission planner will need to be modified a bit to allow mission items to be handled when the drone has disarmed and to not force the drone into standby after landing.

If you'd prefer, the entire `if (autocontinue && is_landed_) { return MISSION_FINISHED; }` part can be removed and then it would just go to standby next iteration with no check of autocontinue.